src/libostree/ostree-core.c

```diff
@@ -333,119 +340,72 @@ _ostree_zlib_file_header_new (GFileInfo         *file_info,
-    {
-      padding_len = alignment - bits;
-      if (!ot_gio_write_update_checksum (output, (guchar*)padding_nuls, padding_len,
-                                         out_bytes_written, checksum,
-                                         cancellable, error))
-        return FALSE;
-    }
-
-  return TRUE;
+  g_autoptr(GVariant) ret = g_variant_new ("(tuuuus@a(ayay))",
+                                           GUINT64_TO_BE (size), GUINT32_TO_BE (uid),
+                                           GUINT32_TO_BE (gid), GUINT32_TO_BE (mode), 0,
+                                           symlink_target, xattrs ?: tmp_xattrs);
+  return variant_to_lenprefixed_buffer (g_variant_ref_sink (ret));
 }
 
-/*
- * _ostree_write_variant_with_size:
- * @output: Stream
- * @variant: A variant
- * @out_bytes_written: (out): Number of bytes written
- * @checksum: (allow-none): If provided, update with written data
- * @cancellable: Cancellable
- * @error: Error
- *
- * Use this function for serializing a chain of 1 or more variants
- * into a stream; the @alignment_offset parameter is used to ensure
- * that each variant begins on an 8-byte alignment so it can be safely
- * accessed.
+/* Serialize a variant to a buffer prefixed with its length. The variant will
+ * have an 8-byte alignment so it can be safely used with `mmap()`.
  */
-gboolean
-_ostree_write_variant_with_size (GOutputStream      *output,
-                                 GVariant           *variant,
-                                 OtChecksum         *checksum,
-                                 GCancellable       *cancellable,
-                                 GError            **error)
+static GBytes *
+variant_to_lenprefixed_buffer (GVariant *variant)
 {
+  /* This string is really a binary memory buffer */
+  g_autoptr(GString) buf = g_string_new (NULL);
   /* Write variant size */
   const guint64 variant_size = g_variant_get_size (variant);
   g_assert (variant_size < G_MAXUINT32);
   const guint32 variant_size_u32_be = GUINT32_TO_BE((guint32) variant_size);
 
-  if (!ot_gio_write_update_checksum (output, &variant_size_u32_be, sizeof (variant_size_u32_be),
-                                     NULL, checksum, cancellable, error))
-    return FALSE;
+  g_string_append_len (buf, (char*)&variant_size_u32_be, sizeof (variant_size_u32_be));
   const gsize alignment_offset = sizeof(variant_size_u32_be);
 
-  /* Pad to offset of 8, write variant */
-  if (!write_padding (output, 8, alignment_offset, NULL, checksum,
-                      cancellable, error))
-    return FALSE;
+  /* Write NULs for alignment. At the moment this is a constant 4 bytes (i.e.
+   * align to 8, since the length is 4 bytes). For now, I decided to keep some
+   * of the (now legacy) more generic logic here in case we want to revive it
+   * later.
+   */
+  const guint alignment = 8;
+  const guchar padding_nuls[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+  guint bits;
+  if (alignment == 8)
+    bits = alignment_offset & 7;
```

Comment:
Not new to this patch, though can we add a comment explaining these? This is essentially a more efficient way to do modulo 8 and 4, right?